src/tool_operate.c

```diff
@@ -315,7 +317,22 @@ static CURLcode pre_transfer(struct GlobalConfig *global,
 
     /* we ignore file size for char/block devices, sockets, etc. */
     if(S_ISREG(fileinfo.st_mode))
+#ifdef __VMS
       uploadfilesize = fileinfo.st_size;
+#else
+    {
+      /* When the upload file is stdin, or when the upload file is
+         /dev/std{in,out,err} or /dev/fd/N on BSDs, the offset may not
+         be 0 */
```

Comment:
can you give an example of this happening

Reply:
Sure!
For stdin as upload file (`-T -`), it is the third example in the PR description: `{ read -r x && src/curl -T - localhost:40400 ;} < ~/.bashrc`.
For an example of `-T /dev/fd/N` etc on BSDs, see #12177.